brain/detail/circuit.h

```diff
@@ -222,12 +189,335 @@ class Circuit::Impl
+
+        Quaternionfs output;
+
+        for (const auto gid : gids)
+        {
+            const float rX = rotationAngleX[gid - startIdx];
+            const float rY = rotationAngleY[gid - startIdx];
+            const float rZ = rotationAngleZ[gid - startIdx];
+
+            const float cX = std::cos(rX), cY = std::cos(rY), cZ = std::cos(rZ);
+            const float sX = std::sin(rX), sY = std::sin(rY), sZ = std::sin(rZ);
+
+            // These are the values given by multiplying the rotation
+            // matrices for R(X)R(Y)R(Z) i.e. extrinsic rotation around Z
+            // then Y then X
+            vmml::Matrix3f mtx;
+
+            mtx(0, 0) = cY * cZ;
+            mtx(0, 1) = -cY * sZ;
+            mtx(0, 2) = sY;
+            mtx(1, 0) = cZ * sX * sY + cX * sZ;
+            mtx(1, 1) = cX * cZ - sX * sY * sZ;
+            mtx(1, 2) = -cY * sX;
+            mtx(2, 0) = sX * sZ - cX * cZ * sY;
+            mtx(2, 1) = cZ * sX + cX * sY * sZ;
+            mtx(2, 2) = cX * cY;
+
+            output.push_back(Quaternionf(mtx));
+        }
+
+        return output;
+    }
+    virtual Strings getMorphologyNames(const GIDSet& gids) const
+    {
+        if (gids.empty())
+            return Strings();
+
+        const size_t startIdx = *gids.begin();
+        const size_t endIdx = *gids.rbegin() + 1;
+
+        const auto names =
+            nodeGroup.getAttribute<std::string>("morphology_file", startIdx,
+                                                endIdx);
+        Strings output;
+        for (const auto gid : gids)
+            output.push_back(names[gid - startIdx]);
+
+        return output;
+    }
+
+    virtual URI getMorphologySource() const final { return morphologySource; }
+    virtual const brion::URI& getCircuitSource() const
+    {
+        LBUNIMPLEMENTED;
+        return *(URI*)(delete_this);
+    }
+    virtual void saveMorphologyToCache(
+        const std::string& /*uri*/, const std::string& /*hash*/,
+        neuron::MorphologyPtr /*morphology*/) const
+    {
+        LBUNIMPLEMENTED;
+        return;
+    }
+
+    virtual CachedMorphologies loadMorphologiesFromCache(
+        const std::set<std::string>& /*hashes*/) const
+    {
+        LBUNIMPLEMENTED;
+        return CachedMorphologies();
+    }
+    virtual void saveSynapsePositionsToCache(
+        const uint32_t /*gid*/, const std::string& /*hash*/,
+        const brion::SynapseMatrix& /*value*/) const
+    {
+        LBUNIMPLEMENTED;
+        return;
+    }
+    virtual CachedSynapses loadSynapsePositionsFromCache(
+        const Strings& /*keys*/) const
+    {
+        LBUNIMPLEMENTED;
+        return CachedSynapses();
+    }
+    virtual void _findSynapsePositionsColumns() const
+    {
+        LBUNIMPLEMENTED;
+        return;
+    }
+    virtual const brion::SynapseSummary& getSynapseSummary() const
+    {
+        LBUNIMPLEMENTED;
+        return *(brion::SynapseSummary*)(delete_this);
```

Comment:
Use a static variable for all these return values.

Reply:
I cannot use a variable for it since Synapse constructor will throw an exception if I try to open a nonexistent file.

Reply:
Yes, that is what I will do.